internal/k8s/statusupdater.go

```diff
@@ -82,9 +83,10 @@ func (suh *StatusUpdateHandler) Start(stop <-chan struct{}) error {
 			suh.Log.WithField("name", upd.NamespacedName.Name).
 				WithField("namespace", upd.NamespacedName.Namespace).
 				Debug("received a status update")
-			uObj, err := suh.Clients.DynamicClient().
-				Resource(upd.Resource).
-				Namespace(upd.NamespacedName.Namespace).Get(context.TODO(), upd.NamespacedName.Name, metav1.GetOptions{})
+
+			// Fetch the lister cache for the informer associated with this resource.
+			lister := suh.InformerFactory.ForResource(upd.Resource).Lister()
+			uObj, err := lister.ByNamespace(upd.NamespacedName.Namespace).Get(upd.NamespacedName.Name)
```

Comment:
We're not explicitly waiting for cache sync here before making the Get call; I'm not sure yet if that's actually an issue or not, as it's possible that this code can't actually execute until other sections that *do* wait for cache sync have run. Looking into it some more.

Reply:
I guess it's probably OK since this is all getting triggered off an event handler for the HTTPProxy informer anyway.

Reply:
Based on slack threads, my understanding is that the only way things would land in the DAG is via informer events.

So while I think you are right (based on Contour's architecture) that the informer cache may not have fully synchronized, I'd suppose that the resource in question would be present in the cache (since informer events announcing its presence in the cache are what triggered its presence in the DAG).

If the cache is stale, then I'd expect an optimistic concurrency failure on the UpdateStatus.  If that proves to be a problem, then we can wrap this fetch/mutate/update in a retry loop that uses the client on subsequent retries.

Reply:
Yep, that sounds right/reasonable to me!

Reply:
It's notable that one of the ways more conventional controllers handle this is to decouple the pre-sync informer events from the post-sync reconciliation using the workqueue (cc @youngnick ).  Basically:

1. Set up informer events that dump into workqueue
2. Wait for sync
3. Continuously drain workqueue

The direct attachment of the event and the processing makes waiting for informer syncs before leverage the cache a trickier proposition.

I bring it up because @youngnick and I were chatting about workqueue usage last night, and this felt like a relevant use case in the broader ecosystem.